Olivierg/gateway usage gauge

ogaca-dd opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This PR adds a new gauge metric `datadog.otel.gateway`. This metric is set at `1` when the collector is used as a gateway otherwise, the value is `0`.
If `datadog.host.name` was sent with two different values, then the collector is used as a gateway.

Note: [The PR on opentelemetry-mapping-go](https://github.com/DataDog/opentelemetry-mapping-go/pull/479) should be merged before this PR and `go.mod` must use a commit on main instead of `0.24.1-0.20250124131402-bcd6fd1617a6`

Associated PRs:
- https://github.com/DataDog/opentelemetry-mapping-go/pull/479
- https://github.com/DataDog/datadog-agent/pull/33370
<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes

<!--Describe what testing was performed and which tests were added.-->
#### Testing
The following commands generates the metric `datadog.otel.gateway` at `1` 
```
telemetrygen metrics --otlp-insecure   --otlp-attributes "datadog.host.name"=\"host1\"
telemetrygen metrics --otlp-insecure   --otlp-attributes "datadog.host.name"=\"host2\"
```

Same tests were performed for `traces` and for `logs` when `isLogsAgentExporterEnabled` is false.


<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

ogaca-dd committed:
Send datadog.otel.gateway for metrics and traces

ogaca-dd committed:
Update dependencies

ogaca-dd committed:
Send datadog.otel.gateway for logs when isLogsAgentExporterEnabled is false

songy23 commented on exporter/datadogexporter/go.mod:
```diff
@@ -23,18 +23,18 @@ require (
-	github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes v0.24.0
-	github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/logs v0.24.0
-	github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/metrics v0.24.0
+	github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata v0.24.1-0.20250124131402-bcd6fd1617a6
```
Please make a new release of mapping-go after your PR is merged

songy23 commented on exporter/datadogexporter/internal/metrics/series.go:
```diff
@@ -67,6 +69,17 @@ func DefaultMetrics(exporterType string, hostname string, timestamp uint64, tags
 
+// GatewayUsageGauge creates a gauge metric to report is there is a gateway
+func GatewayUsageGauge(timestamp uint64, hostname string, tags []string, gatewayUsage *attributes.GatewayUsage) datadogV2.MetricSeries {
+	series := NewGauge("datadog.otel.gateway", timestamp, gatewayUsage.Gauge(), tags)
```
What if the gateway host is from traces, metrics and logs, does this gauge report value of 3?

ogaca-dd replied:
`gatewayUsage.Gauge()` returns either `1` or `0` so in your case the value is `1`.

songy23 reviewed (commented):
This needs a changelog too

ogaca-dd committed:
Update opentelemetry-mapping-go to v0.25.0

ogaca-dd committed:
Update github.com/DataDog/datadog-agent to commit in main

ogaca-dd committed:
Merge branch 'main' of ssh://github.com/open-telemetry/opentelemetry-collector-contrib into olivierg/gateway-usage-gauge

ogaca-dd committed:
Add changelog

ogaca-dd commented:
> This needs a changelog too
Changelog added.

ogaca-dd committed:
Fix github.com/DataDog/datadog-agent/comp/otelcol/otlp/components/statsprocessor version

ogaca-dd committed:
Merge branch 'main' of ssh://github.com/open-telemetry/opentelemetry-collector-contrib into olivierg/gateway-usage-gauge

ogaca-dd committed:
make gotidy

songy23 commented:
more breaking changes: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/13159514653/job/36725209514?pr=37499

ogaca-dd committed:
Fix integration tests

ogaca-dd committed:
Make sure gatewayUsage is not null

ogaca-dd committed:
Fix formating

ogaca-dd committed:
Fix test in datadogconnector

songy23 commented:
The logs tests still failed https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/13182084693/job/36795711129?pr=37499

ogaca-dd committed:
Fix "json: cannot unmarshal object into Go value of type testutil.JSONLogs"
TestLogsExporter creates a DatadogLogServerMock which assumes that every request type are "log". As hosts is empty in exportUsageMetrics, sends the series only if it is not empty solve the issue.

ogaca-dd committed:
Fix Test_metricsExporter_PushMetricsData

ogaca-dd committed:
Fix exporter test

ogaca-dd committed:
Fix lint issue in metrics_exporter_test.go

songy23 committed:
Update exporter/datadogexporter/internal/metrics/series.go
